src/libraries/Microsoft.Extensions.Options.DataAnnotations/src/DataAnnotationValidateOptions.cs

```diff
@@ -52,19 +55,86 @@ public ValidateOptionsResult Validate(string? name, TOptions options)
             ThrowHelper.ThrowIfNull(options);
 
             var validationResults = new List<ValidationResult>();
-            if (Validator.TryValidateObject(options, new ValidationContext(options), validationResults, validateAllProperties: true))
+            HashSet<object>? visited = null;
+            List<string>? errors = null;
+
+            if (TryValidateOptions(options, options.GetType().Name, validationResults, ref errors, ref visited))
             {
                 return ValidateOptionsResult.Success;
             }
 
-            string typeName = options.GetType().Name;
-            var errors = new List<string>();
-            foreach (ValidationResult result in validationResults)
+            Debug.Assert(errors is not null && errors.Count > 0);
+
+            return ValidateOptionsResult.Fail(errors);
+        }
+
+        [RequiresUnreferencedCode("This method on this type will walk through all properties of the passed in options object, and its type cannot be " +
+            "statically analyzed so its members may be trimmed.")]
+        private static bool TryValidateOptions(object options, string qualifiedName, List<ValidationResult> results, ref List<string>? errors, ref HashSet<object>? visited)
+        {
+            if (visited is not null && visited.Contains(options))
```

Comment:
Couldn't you also short circuit the method if `options` is null as well?

Reply:
`TryValidateOptions` never called with null. We always check for null before we call it. I'll add assert for null to ensure no one in the future call it with null.